rclpy/src/rclpy/_rclpy.c

```diff
@@ -3213,7 +3241,53 @@ rclpy_take(PyObject * Py_UNUSED(self), PyObject * args)
 
   // if take failed, just do nothing
   destroy_ros_message(taken_msg);
-  Py_RETURN_NONE;
+  return NULL;
+}
+
+/// Take a message from a given subscription
+/**
+ * \param[in] pysubscription Capsule pointing to the subscription to process the message
+ * \param[in] pymsg_type Instance of the message type to take
+ * \return Python message with all fields populated with received message
+ */
+static PyObject *
+rclpy_take(PyObject * self, PyObject * args)
+{
+  PyObject * result = __rclpy_take_with_info(self, args, NULL);
+  if (result != NULL) {
+    return result;
+  } else {
+    Py_RETURN_NONE;
+  }
+}
+/// Take a message and its message_info from a given subscription
+/**
+ * \param[in] pysubscription Capsule pointing to the subscription to process the message
+ * \param[in] pymsg_type Instance of the message type to take
+ * \return Tuple of (Python message with all fields populated with received message, message_info)
+ */
+static PyObject *
+rclpy_take_with_info(PyObject * self, PyObject * args)
+{
+  rmw_message_info_t message_info;
+  PyObject * msg = __rclpy_take_with_info(self, args, &message_info);
+  if (msg == NULL) {
+    Py_RETURN_NONE;
+  }
+
+  PyObject * mi_dict = __rclpy_message_info_to_dict(&message_info);
+  if (mi_dict == NULL) {
+    Py_DECREF(msg);
+    Py_RETURN_NONE;
+  }
+  PyObject * rt = PyTuple_Pack(2, msg, mi_dict);
+  if (rt == NULL) {
+    Py_DECREF(msg);
+    Py_DECREF(mi_dict);
+    Py_RETURN_NONE;
+  }
+
+  return rt;
```

Comment:
Do we need to decref msg and mi_dict since it is owned by the tuple now too? I actually don't know.

Reply:
The docs for `PyTuple_Pack` say it "behaves like PyBuildValue(OO)" which increments the reference count. So yes, we should. Done.